simoc_server/agent_model/agents/structure.py

```diff
@@ -128,22 +134,97 @@ class Greenhouse(Structure):
         pass
-    def check_agents(self):
-        pass
 
+    def place_plant_inside(self, agent):
+        self.plants.append(agent)
+
+    def remove_plant(self, agent):
+        self.plants.remove(agent)
+
+#Harvester 
+
+class Harvester(EnclosedAgent):
+    agent_type_name = "harvester"
+    # TODO harvester harvests all plants in one step, maybe needs to be incremental
+    # Plant matter densities
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.plant_mass_density = 721 #NOT ACTUAL DENSITY kg/m^3
+
+    def step(self):
+        if (self.structure.plants_ready > 0):
+            self.harvest()
+
+    def harvest(self):
+        for x in self.structure.plants:
+            if(x.status == "grown"):
+                plant_age = timedelta_to_days(self.model.model_time - x.model_time_created)
+                edible_mass = x.get_agent_type_attribute("edible") * plant_age
+                inedible_mass = x.get_agent_type_attribute("inedible") * plant_age
+                #Needs different densities for inedible/edible, add to plant attr
+                self.ship(to_volume(edible_mass, self.plant_mass_density), to_volume(inedible_mass, self.plant_mass_density))
+                self.structure.remove_plant(x)
+                x.destroy()
+                self.structure.plants_ready -= 1
+            if(self.structure.plants_ready == 0):
+                break
+
+    def ship(self, edible, inedible):
+        possible_storage = self.model.get_agents(StorageFacility)
+        edible_to_store = edible
+        inedible_to_store = inedible
+        for x in possible_storage:
+            if(edible_to_store > 0):
+                edible_to_store -= x.store("edible_mass", edible)
+            if(inedible_to_store > 0):
+                inedible_to_store -= x.store("inedible_mass", inedible)
+            if(edible_to_store == 0 and inedible_to_store == 0):
+                break
+
+#Planter
+
+class Planter(EnclosedAgent):
+    agent_type_name = "planter"
+    # TODO right now just grows generic plant, the planter should choose a specific type somehow
+    # TODO planter plants everything in one step, should be incremental
+    # TODO planter should use soil
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def step(self):
+        #FOR TESTING print(self.structure.plants_housed)
+        #FOR TESTING print(self.structure.max_plants)
+
+        if(len(self.structure.plants) < self.structure.max_plants):
+            to_plant = self.structure.max_plants - len(self.structure.plants)
+            self.plant(to_plant) 
+
+        #FOR TESTING print(self.structure.plants[0].status)
+
+    def plant(self, number_to_plant):
+        for x in range(0, number_to_plant):
+            plant_agent = agents.PlantAgent(self.model, structure=self.structure)
+            self.model.add_agent(plant_agent)
+            self.structure.place_agent_inside(plant_agent)
```

Comment:
Duplicate line here. Also I fixed the issue where the agent is not added to the structure passed in by the constructor. The one is my bad.